## Segment 3 - Generating summary statistics using pandas and scipy

In [29]:
import numpy as np
import pandas as pd
from pandas import Series, DataFrame

import scipy
from scipy import stats

In [30]:
from google.colab import files
uploaded = files.upload()

In [31]:
cars = pd.read_csv('mtcars.csv', index_col=0)

#cars.columns = ['car_names','mpg','cyl','disp','hp','drat','wt','qsec','vs','am','gear','carb']

cars.head()

mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  \
Mazda RX4          21.0    6  160.0  110  3.90  2.620  16.46   0   1     4   
Mazda RX4 Wag      21.0    6  160.0  110  3.90  2.875  17.02   0   1     4   
Datsun 710         22.8    4  108.0   93  3.85  2.320  18.61   1   1     4   
Hornet 4 Drive     21.4    6  258.0  110  3.08  3.215  19.44   1   0     3   
Hornet Sportabout  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3   

                   carb  
Mazda RX4             4  
Mazda RX4 Wag         4  
Datsun 710            1  
Hornet 4 Drive        1  
Hornet Sportabout     2

### Looking at summary statistics that decribe a variable's numeric values

In [32]:
cars.sum()

mpg      642.900
cyl      198.000
disp    7383.100
hp      4694.000
drat     115.090
wt       102.952
qsec     571.160
vs        14.000
am        13.000
gear     118.000
carb      90.000
dtype: float64

In [33]:
numerical_cols = cars.columns
cars[numerical_cols].mean()

mpg      20.090625
cyl       6.187500
disp    230.721875
hp      146.687500
drat      3.596563
wt        3.217250
qsec     17.848750
vs        0.437500
am        0.406250
gear      3.687500
carb      2.812500
dtype: float64

In [34]:
cars[numerical_cols].median()

mpg      19.200
cyl       6.000
disp    196.300
hp      123.000
drat      3.695
wt        3.325
qsec     17.710
vs        0.000
am        0.000
gear      4.000
carb      2.000
dtype: float64

In [35]:
cars.max()

mpg      33.900
cyl       8.000
disp    472.000
hp      335.000
drat      4.930
wt        5.424
qsec     22.900
vs        1.000
am        1.000
gear      5.000
carb      8.000
dtype: float64

In [36]:
mpg = cars.mpg
mpg.idxmax()


'Toyota Corolla'

In [37]:
mpg.idxmin()

'Cadillac Fleetwood'

# question: find the max and min hoursepower car name from our dataset.

In [38]:
max_hp_car = cars['hp'].idxmax()
max_hp = cars['hp'].max()

min_hp_car = cars['hp'].idxmin()
min_hp = cars['hp'].min()

print(f"Car with MAX horsepower: {max_hp_car} ({max_hp} hp)")
print(f"Car with MIN horsepower: {min_hp_car} ({min_hp} hp)")


Car with MAX horsepower: Maserati Bora (335 hp)
Car with MIN horsepower: Honda Civic (52 hp)


### Looking at summary statistics that describe variable distribution

In [39]:
cars.std()

mpg       6.026948
cyl       1.785922
disp    123.938694
hp       68.562868
drat      0.534679
wt        0.978457
qsec      1.786943
vs        0.504016
am        0.498991
gear      0.737804
carb      1.615200
dtype: float64

In [40]:
cars.var()

mpg        36.324103
cyl         3.189516
disp    15360.799829
hp       4700.866935
drat        0.285881
wt          0.957379
qsec        3.193166
vs          0.254032
am          0.248992
gear        0.544355
carb        2.608871
dtype: float64

In [41]:
gear = cars.gear # gear= cars['gear']
gear.value_counts()

gear
3    15
4    12
5     5
Name: count, dtype: int64

In [42]:
cars.describe()

mpg        cyl        disp          hp       drat         wt  \
count  32.000000  32.000000   32.000000   32.000000  32.000000  32.000000   
mean   20.090625   6.187500  230.721875  146.687500   3.596563   3.217250   
std     6.026948   1.785922  123.938694   68.562868   0.534679   0.978457   
min    10.400000   4.000000   71.100000   52.000000   2.760000   1.513000   
25%    15.425000   4.000000  120.825000   96.500000   3.080000   2.581250   
50%    19.200000   6.000000  196.300000  123.000000   3.695000   3.325000   
75%    22.800000   8.000000  326.000000  180.000000   3.920000   3.610000   
max    33.900000   8.000000  472.000000  335.000000   4.930000   5.424000   

            qsec         vs         am       gear     carb  
count  32.000000  32.000000  32.000000  32.000000  32.0000  
mean   17.848750   0.437500   0.406250   3.687500   2.8125  
std     1.786943   0.504016   0.498991   0.737804   1.6152  
min    14.500000   0.000000   0.000000   3.000000   1.0000  
25%    16.892500   0.000000   0.000000   3.000000   2.0000  
50%    17.710000   0.000000   0.000000   4.000000   2.0000  
75%    18.900000   1.000000   1.000000   4.000000   4.0000  
max    22.900000   1.000000   1.000000   5.000000   8.0000

For above, 25% is percentile calculated using below formula. res_25 = (value falling at 25% index) + ((value falling one after 25% index)-(value falling at 25% index))*(3/4)

So in our case it would be 15.2 + (15.5 - 15.2) * (3/4) = 15.425

Similarly, it calcuates 50% as below : res_50 = (value falling at 50% index) + ((value falling one after 50% index)-(value falling at 50% index))*(1/2)

and for 75% res_75 = (value falling at 75% index) + ((value falling one after 75% index)-(value falling at 75% index))*(1/4)

